### ISSUE TITLE
integration with IPE tracker

### ISSUE THREAD

**Issue by cgwalters** [containers/composefs#360]
As of the current Linux kernel merge window, [the IPE LSM](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=a430d95c5efa) was merged. Per [the docs](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/diff/Documentation/admin-guide/LSM/ipe.rst?id=a430d95c5efa) an example policy is e.g. 

`op=EXECUTE dmverity_signature=TRUE action=ALLOW`

This relies on the idea that the dm-verity signature was verified in-kernel I believe. In contrast, the fsverity bits in IPE match by digest (annoying at scale) or rely on the in-kernel fsverity signatures which the fsverity maintainer doesn't like.

As folks here know with composefs we can combine the advantages of these. I think there is a bit of plumbing work that needs to be done, Eric Biggers (fsverity maintainer) noted this at one point. 

I think in the end state what we want is something like:

`op=EXECUTE composefs_signature=TRUE action=ALLOW`

Now, this gets into a messy topic of whether signatures are verified in kernel or in user space, and if in user space how we set a property that then later is trusted by the kernel.

And since composefs doesn't exist at the kernel level maybe this would need to be e.g. `op=EXECUTE overlayfs_fsverity_signature=TRUE action=ALLOW` or so (and covering both the loopback-erofs fsverity vs backing file fsverity.

And in general I think we need propagate metadata across our chain of 3 components (overlays, erofs, and the backing filesystem) so the higher level overlayfs inode that the LSM sees can reliably see that e.g. fsverity was required and enabled on the backing file, etc.

**Comment by jluebbe** [comment]
File-backed mount support for EROFS has been [merged for Linux v6.12](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=fb176750266a). As that avoids the loopback device, it should make it easier to associate the fsverity signature to the erofs mount.

**Comment by hsiangkao** [comment]
> File-backed mount support for EROFS has been [merged for Linux v6.12](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=fb176750266a). As that avoids the loopback device, it should make it easier to associate the fsverity signature to the erofs mount.

Geert has requested for marking it as `default n` https://lore.kernel.org/r/CAMuHMdVqa2Mjqtqv0q=uuhBY1EfTaa+X6WkG7E2tEnKXJbTkNg@mail.gmail.com
That was why I tended to avoid this stuff in the beginning (because the kernel file-backed mount controversy) although I know there are many use cases.

**Comment by alexlarsson** [comment]
Yeah, what would be very good is if EROFS, when using the file-backed mounts could take an optional fs-verity digest mount option, and do the validation of it in the kernel. Then it could somehow mark the mount with a "on fs-verity" flag. And then further, overlayfs could, if all layers except data only layers are marked "on fs-verity" and veryity=enabled is supported, then it could also set the "on fs-verity" flag.

Then one could use this flag with something like IPE.

However, this is just the "fs-verity" part, we also would need to propagate the fact that the initial mountpoint was signed, and that seems much more tricky.

**Comment by giuseppe** [comment]
> , if all layers except data only layers are marked "on fs-verity" and veryity=enabled is used, then it could also set the "on fs-verity" flag.

would the order matter?  If two overlay layers are switched, the resulting tree could be different

**Comment by alexlarsson** [comment]
Yeah, that is true, maybe it would be easiest to just require there to be a single signed layer.

**Comment by hsiangkao** [comment]
Hi folks, I don't have deep understanding of IPE but since it's related to EROFS I took into for a while.

I guess the IPE policy of Overlayfs should follow the metadata layer instead of data-only layers, because data-only file blobs are just deduplicated blobs and could be even from several _sources_ or _vendors_: I think we shouldn't apply IPE or some signed-sig on those blob files (since they only provide identified binary data), otherwise one blob file may be shipped by many vendors (so it would got many sigs in principle, I don't think it's useful.)

Specifically, see what `ipe_build_eval_ctx` does, it calls `ino = d_real_inode(file->f_path.dentry);` -> `d_real(dentry, D_REAL_DATA);`,
I'm not sure why it's D_REAL_DATA here.

In principle, IPE will follow the metadata-only layer/fs IPE policy if it's D_REAL_METADATA instead, then we could apply a dm-verity signed sig to the meta-only layer and then a IPE policy like:
`op=EXECUTE dmverity_signature=TRUE action=ALLOW`
will just allow execute files in the signed metadata-layer by dm-verity.  BTW, I followed the related logic, I think we cannot use fsverity here because it's a per-file policy rather than per-device/per-fs policy which can apply to all files in the filesystem.

Does it make sense? Please correct me if I'm wrong.

**Comment by alexlarsson** [comment]
fs-verity IPE rules are not very interesting in general, because any file can be made fs-verity, it doesn't imply any particular trust in them. What is interesting for IPE is external signatures where someone declared that "this particular set of files are trusted". For dm-verity that means all the files from any filesystem on that partition. For composefs, it means "the data in the composefs, and any external referenced files that matches the digests encoded in the erofs".

So, yes, the signatures are not on the "backing files" in the data-only layer themselves. Only when you access those via a signed erofs, then they are considered "signed".

I have not really looked at the implementation side of IPE in the kernel, but basically this is how I believe this must work:
 * Somehow we must be able to sign an erofs image. The signature could be embedded in the fs, or supplied separately at mount time. 
 * The signature should not be of the full erofs content, but rather of its fs-verity digest. That way it is efficient to verify in the "mount from file" code. (I.e. we don't have to read the entire erofs image file.)
 * When you mount such an EROFS it tries to verify the signature against a key in the keyring. If it matches, we mark the erofs superblock as "signed". 
 * Optionally we could have a mount option where the mount fails if its not signed, but that is not always what you want.
 * If you mount an overlayfs, we check if the lowest non-data-only layer is a signed erofs superblock, and the "verity=require" overlayfs mount option is used. If this is true, then all files from the lowest layers are considered "signed", including those read from any data-only layer (because they are required to have a verity digest which will be verified on access)
 * Files from higher layers (if any) are not considered signed.

Open question:
* If there are multiple signed erofs lower layers, do we still trust them? One could remount and reorder such layers which would let you access files that otherwise would not be accessible, but you can't get files with arbitrary data this way,
* How does overlayfs work with dm-verity signatures today? If I mount a signed dm-verity filesystem and mount a lower layer from it, are those files then considered signed?

**Comment by hsiangkao** [comment]
...

> * When you mount such an EROFS it tries to verify the signature against a key in the keyring. If it matches, we mark the erofs superblock as "signed".
> * Optionally we could have a mount option where the mount fails if its not signed, but that is not always what you want.
> * If you mount an overlayfs, we check if the lowest non-data-only layer is a signed erofs superblock, and the "verity=require" overlayfs mount option is used. If this is true, then all files from the lowest layers are considered "signed", including those read from any data-only layer (because they are required to have a verity digest which will be verified on access)
> * Files from higher layers (if any) are not considered signed.

Personally I don't think those steps are needed if IPE uses `d_real(inode, D_REAL_METADATA)`, because IPE policies (whether or not to execute) will be applied to the "real" metadata inode, in other words, EROFS itself inode itself.

Overlayfs just check fsverity digest according to the EROFS xattrs as previously, yes,  "verity=require" is a must to keep the trust of chain works.

But I might be wrong on this, I don't think overlayfs needs to get a "signed" mark or even some additional feature. Just make IPE use the real metadata inode (D_REAL_METADATA) for IPE policies instead of EXT4/XFS data-only inodes (D_REAL_DATA).

> 
> Open question:
> 
> * If there are multiple signed erofs lower layers, do we still trust them? One could remount and reorder such layers which would let you access files that otherwise would not be accessible, but you can't get files with arbitrary data this way,

Does it really matter if "verity=require" (I think it's a must), if the digest mismatches, it will fail the I/O as the current status even IPE policies pass.

> * How does overlayfs work with dm-verity signatures today? If I mount a signed dm-verity filesystem and mount a lower layer from it, are those files then considered signed?

According to the IPE codebase, I think dm-verity is a per-filesystem signature, also see `ipe_build_eval_ctx()` and `evaluate_dmv_sig_true` because it uses the blockdev context `ctx->ipe_bdev = ipe_bdev(INO_BLOCK_DEV(inode))` (basically derived from `s_bdev`) rather than fsverity the inode itself.

**Comment by alexlarsson** [comment]
I never looked at the actual IPE implementation, but if you're right it seems that this would not be so much work.

>>If there are multiple signed erofs lower layers, do we still trust them? One could remount and reorder such layers which would let you access files that otherwise would not be accessible, but you can't get files with arbitrary data this way,

>Does it really matter if "verity=require" (I think it's a must), if the digest mismatches, it will fail the I/O as the current status even IPE policies pass.

My worry isn't really one of matching verity digest, but of access in the first place. For example, suppose your base for these things is layered like an OCI image, and you have a CVE in /bin/app. To fix this you push out a fixed version of your image, which has a new layer on top that replaces /bin/app. If the layers are mounted as expected you can't see the old /bin/app, but if you reorder them you can. So, signing the individual layers is less expressive than signing the combined order.

Otoh, the same would be true for a dm-verity update if the old dm-verity image was still around and they were signed with the same key. So, maybe this isn't a huge deal.

**Comment by giuseppe** [comment]
> I think it could be just identified with the new overlayfs data-only layer feature, since only composefs use cases use this, see: `Documentation/filesystems/overlayfs.rst` and [http://github.com/torvalds/linux/commit/37ebf056d6cf](https://github.com/torvalds/linux/commit/37ebf056d6cf).

would it be better to use `verity=require`?  `overlay_require=TRUE`?

**Comment by hsiangkao** [comment]
> > I think it could be just identified with the new overlayfs data-only layer feature, since only composefs use cases use this, see: `Documentation/filesystems/overlayfs.rst` and [http://github.com/torvalds/linux/commit/37ebf056d6cf](https://github.com/torvalds/linux/commit/37ebf056d6cf).
> 
> would it be better to use `verity=require`? `overlay_require=TRUE`?

Yeah, that is fine too, anyway I think it can be clearly seperated from generic overlayfs mounts.

**Comment by cgwalters** [comment]
> Somehow we must be able to sign an erofs image. The signature could be embedded in the fs, or supplied separately at mount time.

https://github.com/containers/composefs-rs/pull/224/ covers this for the OCI path I think.

**Comment by jxwufan** [comment]
I almost got a poc working, who should I cc for composefs?

**Comment by cgwalters** [comment]
Hi @jxwufan - I think here is great, you're in the right place. (Eventually I hope to consolidate this repo with composefs-rs, but for now it's fine)

I forgot to mention here that I spent some LLM tokens on generating https://github.com/cgwalters/rhel-bootc-examples/tree/sealing-gen2-ipe/sealing-gen2-ipe/kernel-build - but I didn't spend much time trying to clean them up. I am curious how they differ from your PoC code.

**Comment by jxwufan** [comment]
I think the key difference is mine anchor the trust by hash, the ai do provide another way that I never thought of which might be better. FYI my poc requires a new public interface from ovlfs but that might not be desired. Will it be possible to get erofs backing file hash in overlayfs/namei.c? I can post mine in a repo before actually sending to the list.

**Comment by jxwufan** [comment]
https://github.com/jxwufan/composefs-poc

**Comment by cgwalters** [comment]
I think it is a valid use case to anchor by hash.

However, IMO it's *significantly* more ergonomic to just anchor trust in a valid fsverity signature on the erofs and have everything else flow from that - we've validated all the metadata that way, and the metadata includes fsverity digests. We don't want to have to include a signature on each individual backing file - again the idea is trust flows from that one signed bit of metadata.

**Comment by jxwufan** [comment]
Agreed, however fsverity signature can be problematic as the fsverity maintainer strongly discouraging using it, I would suggest if anchoring root hash is acceptable it should be the first step.

**Comment by cgwalters** [comment]
Yeah, we have had that debate about fsverity signatures in other threads.

I think the hard part is chaining "trust in fsverity digest of erofs" to the underlying files/objects. How trust in the fsverity digest of the erofs is determined is seems very straightforward - an external specification of the digest or fsverity signatures are already implemented by IPE.

BTW on the naming...conceptually "composefs" doesn't exist at the kernel level, and so perhaps we should call it `overlayfs_fsverity_require_signature=TRUE` and `overlayfs_fsverity_require_digest=` at the IPE level? In theory we could back the intermediate chain with something other than EROFS (not that anyone has any plans to, no reason). But I think conceptually the actually important thing here is the overlayfs chaining through to the backing files, do you agree?

**Comment by jxwufan** [comment]
getting a good name while making it concise is challenging, the ovlfs+fsverity name can be misleading as it may be read as fsverity files in a overlayfs, i'm short of ideas right now, and any suggestions are welcome

**Comment by hsiangkao** [comment]
@jxwufan just a quick glance since I don't have enough time to look over it now,

```diff
+struct dentry *ovl_get_verity_composefs_metadata_dentry(struct dentry *dentry)
+{
+	if (!dentry || !d_is_reg(dentry))
+		return NULL;
+
+	while (dentry && dentry->d_sb->s_type == &ovl_fs_type) {
```
I still wonder why d_real(D_REAL_METADATA) cannot be directly used (I don't see the strong reason in the comment above), since:
 1) it's quite hard to export a symbol from an endpoint kernel module like overlayfs;
 2)`dentry->d_sb->s_type == &ovl_fs_type` seems a no-go at least for me, but I don't know overlayfs guys' opinion.

**Comment by jxwufan** [comment]
d_real(D_REAL_METADATA) does not guarantee verity=require, if verity=no it can still get the erofs dentry and we have no way to know is there a fsverity check happening. Now I do see another way, we can tag the data inode when overlayfs did a fsverity hash check. However I'm not quite sure the behavior difference between verity=yes and verity=require. And could there be any race conditions happen between d_real(D_REAL_METADATA) and d_real(D_REAL_DATA)?

Update: I think tagging the data inode won't work. So if we tag a data inode a boolean say it is verified by a ovlfs or not. An attacker could create a valid composefs with verity=required to tag the data inode first. Then create another ovlfs with verity=no and ipe has no way to tell whether the second ovlfs can be trusted.

**Comment by hsiangkao** [comment]
>  And could there be any race conditions happen between d_real(D_REAL_METADATA) and d_real(D_REAL_DATA)?

I have no idea why D_REAL_DATA is needed, maybe just because I don't know how do you tag the data inode.

> 
> Update: I think tagging the data inode won't work. So if we tag a data inode a boolean say it is verified by a ovlfs or not. An attacker could create a valid composefs with verity=required to tag the data inode first. Then create another ovlfs with verity=no and ipe has no way to tell whether the second ovlfs can be trusted.

Not sure how do you tag the data inode, maybe some skeleton code can explain your idea more; but one simple thought (maybe there are cleaner solutions) is that just introduce another `D_REAL_METADATA_VERIFIED` for example and implement it in ovl_d_real so that only fs-verity verified dentry can be returned, does that work?

**Comment by jxwufan** [comment]
> > And could there be any race conditions happen between d_real(D_REAL_METADATA) and d_real(D_REAL_DATA)?
> 
> I have no idea why D_REAL_DATA is needed, maybe just because I don't know how do you tag the data inode.
> 

Main reason is we need to prevent TOCTOU, we can only know a meta data inode is used to verify a certain inode, however we need to make sure it verified the data inode in use during actual read. 

> > Update: I think tagging the data inode won't work. So if we tag a data inode a boolean say it is verified by a ovlfs or not. An attacker could create a valid composefs with verity=required to tag the data inode first. Then create another ovlfs with verity=no and ipe has no way to tell whether the second ovlfs can be trusted.
> 
> Not sure how do you tag the data inode, maybe some skeleton code can explain your idea more; but one simple thought (maybe there are cleaner solutions) is that just introduce another `D_REAL_METADATA_VERIFIED` for example and implement it in ovl_d_real so that only fs-verity verified dentry can be returned, does that work?

I got another idea, we can tag the meta data inode with the data inode, like save the data inode address in the meta data inode's security blob. However, there is a case this might not work. For example, if an erofs is shared between two ovlfs's with two separate data store, one ovlfs tag could break another. However, there is no valid use case for this scenario I think. 

Update: I just realize saving the address could lead to UAF. Adding a new flag like `D_REAL_METADATA_VERIFIED` could be the best way.

**Comment by cgwalters** [comment]
> For example, if an erofs is shared between two ovlfs's with two separate data store, one ovlfs tag could break another. However, there is no valid use case for this scenario I think.

Today the way userspace tooling mounts a composefs, there are no userspace references to the EROFS, it's just the overlay holding it open. If it was helpful for this, I think we could perhaps even consider erroring out with verity=require on if there are multiple references to the uppermost lower?

> Now I do see another way, we can tag the data inode when overlayfs did a fsverity hash check. However I'm not quite sure the behavior difference between verity=yes and verity=require. 

It's basically whether all files have to have fsverity enabled. I think the use case for `verity=yes` is...probably pretty obscure. For the purposes here I think we can ignore it and only care about IPE with verity=require.

**Comment by hsiangkao** [comment]
> Adding a new flag like `D_REAL_METADATA_VERIFIED` could be the best way.

Without deep thinking, yet I think you could use this way as the first formal mailing list version and see if there are better solutions or better abstraction from other community guys (I think it's far better than exporting overlayfs symbols).

**Comment by andrewdunndev** [comment]
I've been poking at this from the bootc side for a homelab, and ran a few things on stock Fedora 44 (kernel 7.2.4, `ipe` active in the LSM list). Adding them in case they're useful.

On the two identification signals suggested above, the data-only layer looks like the discriminating one. A plain overlayfs mount with a single lower directory and no data-only layer accepts `verity=require` and mounts clean:

```
mount -t overlay overlay -o lowerdir=L,upperdir=U,workdir=W,verity=require /mnt
```

Source string `overlay`, no `::` in `lowerdir`. A composefs mount beside it:

```
overlay composefs:6fd9d5fc...cd3094d ro,lowerdir=/proc/self/fd/10::/proc/self/fd/11,redirect_dir=on,metacopy=on,verity=require
```

Two things fell out of that. The digest a policy would want is already in the mount source, and `fsverity measure` on the EROFS image returns the same value. And Fedora sets `CONFIG_IPE_PROP_FS_VERITY=y` but not `CONFIG_FS_VERITY_BUILTIN_SIGNATURES`, so `fsverity_signature=TRUE` cannot match there at all, which may matter for the signature versus root hash question.

One question: will the property apply to any op, or start with EXECUTE? `op=KMODULE` is where it would change the most for me. I run a sealed bootc homelab carrying out-of-tree ZFS and NVIDIA drivers, so it needs `module.sig_enforce=1` and its own signing key purely because nothing consults the composefs digest when a module loads (https://gitlab.com/dunn.dev/cairn/base/-/blob/main/scripts/assert-sealed.sh checks that state at boot). Those loads do reach IPE with a file, `finit_module` with `MODULE_INIT_COMPRESSED_FILE`, even though Fedora ships modules xz-compressed.

Happy to run a `D_REAL_METADATA_VERIFIED` RFC against that setup if it would help.

**Comment by jxwufan** [comment]
@andrewdunndev Yes, you can use op=KMODULE for signed kernel modules, but you might also need to disable the existing ephemeral kernel signature verification to let the out-of-tree module work. For kernel module loading, only possible blocker is in some cases kmod will decompress the kernel module in userspace and pass it to kernel via a buffer, for that scenario IPE will block the module loading since there is no way to find the module provenance from a buffer. 

verity=require is a required config but the trust is ultimately originated from the erofs backing file. 

CONFIG_FS_VERITY_BUILTIN_SIGNATURES is not really necessary and I'm planning to only support dm-verity signature for composefs. Likely the backing file needs to be from a signed dmverity. I have been kind of busy with other projects, but I do have another rfc for this and will try to see if I can send that out this week.
